doc/bakugo_spec.md

```diff
@@ -0,0 +1,1111 @@
+```
+
+## Letters and digits
+The underscore character `_` is considered a lowercase letter.
+```
+letter        = letter | "_" .
+decimal_digit = "0" … "9" .
+```
+# Lexical elements
+
+## Comments
+
+Comments serve as program documentation. There are two forms:
+
+1. _Line comments_ start with the character sequence `//` and stop at the end of the line.
+2. _General comments_ start with the character sequence `/*` and stop with the first
+subsequent character sequence `*/`.
+
+A comment cannot start inside a _rune_ or _string_ literal, or inside a comment. A general
+comment containing no newlines acts like a space. Any other comment acts like a newline.
+
+## Tokens
+> ***TODO:*** use ASCII values for _space_, _horizontal tabs_, _carriage returns_, _semicolons_ and _newlines_.
+
+Tokens form the vocabulary of the Go language. There are four classes: _identifiers_,
+_keywords_, _operators_, _punctuation_ and _literals_. _White space_, formed from 
+spaces(` `), horizontal tabs(`\t`), carriage returns,(`\r`) and newlines(`\n`), is ignored
+except as it separates tokens that would otherwise combine into a single token. Also, a
+newline or end of file may trigger the insertion of a semicolon(`;`). While breaking the
+input into tokens, the next token is the longest sequence of characters that form a
+valid token.
+
+## Semicolons
+> ***NOTE:*** floating-point, rune or string literal, keywords `break`, `continue` and `fallthrough`, and punctuation `]` are not yet supported.
+
+The formal syntax uses semicolons `";"` as terminators in a number of productions. Go programs may omit most of these semicolons using the following two rules:
+1. When the input is broken into tokens, a semicolon is automatically inserted into the token stream immediately after a line's final token if that token is
+    - an identifier
+    - an integer, floating-point, rune, or string literal
+    - one of the keywords `break`, `continue`, `fallthrough`, or `return`
+    - one of the operators and punctuation `++`, `--`, `)`, `]`, or `}`
+
+2. To allow complex statements to occupy a single line, a semicolon may be omitted before a closing `")"` or `"}"`.
+
+## Identifier
+
+Identifiers name program entities such as variables and types. An identifier is a sequence of one or more letters and digits. The first character in an 
+identifier must be a letter.
+
+```
+identifier = letter { letter | unicode_digit } .
+```
+examples of _identifier_
+```
+a
+_x9
+ThisVariableIsExported
+this_is_Bakugo
+```
+
+## Keywords
+
+The following keywords are reserved and may not be used as _identifiers_.
+```
+func  const  if  type  return  var  else
+```
+</details>
+<details><summary>unsupported keywords</summary>
+<p>
+
+```
+break        default      interface    select       case
+defer        go           map          struct       chan
+goto         package      switch       fallthrough  range
+continue     for          import
```

Comment:
Since we're supporting `if`, `for` shouldn't be too much work beyond that (except for `break`, `continue` and range for). So should we look at including `for` too?

Reply:
Oh, I think we can defer `For statements` for the next stage. As of now, recursion should do the job. `For statements` can be more effective when we have iterable data structures.

Reply:
Sure, makes sense